src/Interpreters/Context.cpp

```diff
@@ -2702,8 +2702,11 @@ void Context::setCurrentQueryId(const String & query_id)
 
     client_info.current_query_id = query_id_to_set;
 
-    if (client_info.query_kind == ClientInfo::QueryKind::INITIAL_QUERY)
+    if (client_info.query_kind == ClientInfo::QueryKind::INITIAL_QUERY
+        && (getApplicationType() != ApplicationType::SERVER || client_info.initial_query_id.empty()))
```

Comment:
According to the comment in ClientInfo.h
when query_kind == INITIAL_QUERY 
initial_query_id is equal to current.
Does not it contradict with the condition?

Reply:
Technically it must be some new kind, something like `REMOTE_INITIAL_QUERY`, but this breaks backward compatibility on protocol level.